ra/ra.go

```diff
@@ -835,7 +860,16 @@ func (ra *RegistrationAuthorityImpl) FinalizeOrder(ctx context.Context, req *rap
 	// Update the order to be status processing - we issue synchronously at the
 	// present time so this is somewhat artificial/unnecessary but allows planning
 	// for the future.
+	//
+	// NOTE(@cpu): After this point any errors that are encountered must update
+	// the state of the order to invalid by setting the order's error field.
+	// Otherwise the order will be "stuck" in processing state. It can not be
+	// finalized because it isn't pending, but we aren't going to process it
+	// further because we already did and encountered an error.
 	if err := ra.SA.SetOrderProcessing(ctx, order); err != nil {
+		// Fail the order with a server internal error - we weren't able to set the
+		// status to processing and that's unexpected & weird.
+		ra.failOrder(ctx, order, probs.ServerInternal("Error setting order processing"))
```

Comment:
I'm not sure we need to make the order unusable here, the failure may be for a variety of database-y reasons that are completely unrelated (i.e. timeouts or connection errors) and everything in `SetOrderProcessing` is done in a transaction. Is there a reason to prevent continued use of the order and require a new order be created to replace it vs. just informing the user of a 500 and letting them try again?

Reply:
>  Is there a reason to prevent continued use of the order and require a new order be created to replace it vs. just informing the user of a 500 and letting them try again?

Yes - as you mentioned in your [other comment](https://github.com/letsencrypt/boulder/pull/3404#discussion_r165498906) the order will be `Status: Processing` when an error occurs in finalization. It won't be retryable because `FinalizeOrder` rejects orders that are not `Status: Pending`. 

When there's an internal error during finalization we could move the order _back_ to `Status: Pending` from `Status: Processing` but the spec doesn't really talk about this transition, I was nervous about second layer consequences of that decision, and it seemed at the time better to just reflect that an error occurred by putting the order into `Status: invalid`.

I think if we want to make orders retryable for finalization errors we should do it as a follow-up issue after a discussion (and maybe a ACME PR/WG thread).

Reply:
> I think if we want to make orders retryable for finalization errors we should do it as a follow-up issue after a discussion (and maybe a ACME PR/WG thread).

👍